### ISSUE TITLE
How do I construct a `__m128i` from 64-bit or 128-bit Rust types?

### ISSUE THREAD

**Issue by shepmaster** [rust-lang-nursery/stdsimd#358]
`_mm_set_epi32` takes standard Rust `i32` values:

```rust
pub unsafe fn _mm_set_epi32(e3: i32, e2: i32, e1: i32, e0: i32) -> __m128i
```

On the other hand, `_mm_set_epi64` takes `__m64` types:

```rust
pub unsafe fn _mm_set_epi64(e1: __m64, e0: __m64) -> __m128i
```

I could do some bit shifting and strongarm my `u64`s into the right 32-bit sized types, but it seems like it would be nicer if I didn't *need* to.

It would also be lovely if this could implement `From` for `u128`&`i128` or even `(i64, i64)` & `(u64, u64)`.

**Comment by alexcrichton** [comment]
The [`_mm_set_epi64x`](https://rust-lang-nursery.github.io/stdsimd/x86_64/stdsimd/arch/x86_64/fn._mm_set_epi64x.html) function, only available on x86_64, should work for 64-bit integers but right now there's no native support for 128-bit integers (may want to ping intel for that!)

In the meantime it seems reasonable to me to provide constructors for the portable types (aka `u64x2`) but we probably won't be providing conveniences for `__m128i` or other vendor types (although they can be bitcasted from the portable `u64x2` types)

**Comment by shepmaster** [comment]
> The `_mm_set_epi64x` function

Hmm. I did see that in the doc search, but this threw me off:

> from highest to lowest.

I assume that was in *contrast* to `_mm_set_epi64` (which would have set from lowest to highest, I guess?). Perhaps that level of detail doesn't belong in the short doc, but instead should be moved to the longer paragraphs? Then the two methods short documentation could be clarified to highlight differences...

**Comment by alexcrichton** [comment]
Heh yeah we could certainly clarify I think better but when in doubt Intel's always got good docs:

* [`_mm_set_epi64x`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_set_epi64x&expand=4614,4614)
* [`_mm_set_epi64`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_set_epi64&expand=4614,4614,4612)

**Comment by shepmaster** [comment]
> although they can be bitcasted

Are there functions for such, or are we talking `transmute` here?

**Comment by shepmaster** [comment]
> Intel's always got good docs:

Perhaps we should also link those from all the intrinsics?

**Comment by gnzlbg** [comment]
@shepmaster 

If you can use `std::simd::u64x2` then you should be able to: 

```rust
let x: __m128i = u64x2::new(1, 2).into_bits();
```

---

@alexcrichton I could implement `From/IntoBits` for  `{i,u}{128}` to the portable vector types, to allow:

```rust
let x: __m128i = u64x2::from_bits(1_i128).into_bits();
```

**Comment by alexcrichton** [comment]
@gnzlbg something like `FromBits<u128> for u64x2` sounds plausible to me!

@shepmaster yeah the 💯 thing to do would for sure add links to all functions to Intel's docs :)

**Comment by shepmaster** [comment]
> If you can use std::simd::u64x2 then you should be able to:

Thanks! However, I'm looking at going back the _other_ way. `_mm_cmpestrm` returns an `__m128i` and I want to pull out the lower 64 bits. I don't see any cross-mentions of either type on their doc pages, nor do I see a `FromBits` or `IntoBits` trait; perhaps I've grossly misunderstood something?

**Comment by shepmaster** [comment]
> perhaps I've grossly misunderstood something?

Ah, I see this trait [in your documentation](https://rust-lang-nursery.github.io/stdsimd/x86_64/stdsimd/simd/trait.FromBits.html), but not in [the nightly docs](https://doc.rust-lang.org/nightly/std/). Is it just that new?

**Comment by alexcrichton** [comment]
@shepmaster ah yeah the `FromBits`/`IntoBits` hasn't made its way into rust-lang/rust just yet, it's only in this repo for the moment.

Extracting values from `__m128i` tends to be pretty hard to do via intrinsics, but in C you'd do something like:

```rust
union A { a: __m128i, b: [u64; 2] }
let a: __m128i = ...;
A { a }.b[0]
```

(or something like that)

**Comment by gnzlbg** [comment]
> Is it just that new?

It's part of the `stdsimd` crate, but might not be part of `std` yet. 

> However, I'm looking at going back the other way.

You have probably figured it out already but to go the other way you just call `into_bits()` again:

```rust
let x: u64x2 = _mm_cmpestrm(...).into_bits();
```

**Comment by shepmaster** [comment]
> How do I construct a `__m128i` from 64-bit or 128-bit Rust types?

TL;DR: 

```rust
let x = _mm_set_epi64x(1, 2);
```

or

```
let x: __m128i = u64x2::new(1, 2).into_bits();
```

**Comment by dragomang87** [comment]
> >    The _mm_set_epi64x function
>
>Hmm. I did see that in the doc search, but this threw me off:
>
> >    from highest to lowest.
>
> I assume that was in contrast to _mm_set_epi64 (which would have set from lowest to highest, I guess?). Perhaps that level of detail doesn't belong in the short doc, but instead should be moved to the longer paragraphs? Then the two methods short documentation could be clarified to highlight differences...

`_mm_set_epi64` is just a legacy version of `_mm_set_epi64x` from when Microsoft introduced 64 bits registers before they got standardized, that's why the argument signature is different: `__m64` for Microsofts original 64 bits and `__int64` for the now standard one (these are the names in Intel's documentation at https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html ). The behaviour is the same, they both set the upper 64 bits with the first argument and the lower 64 bits with the second argument just as you TL;DR'ed:

let e0: i64 = 1;
let e1: i64 = 2;
use std::arch::x86_64::*;
let x: __m128i = unsafe { _mm_set_epi64x(e0, e1) };


> > If you can use std::simd::u64x2 then you should be able to:
> 
> Thanks! However, I'm looking at going back the _other_ way. `_mm_cmpestrm` returns an `__m128i` and I want to pull out the lower 64 bits. I don't see any cross-mentions of either type on their doc pages, nor do I see a `FromBits` or `IntoBits` trait; perhaps I've grossly misunderstood something?

To extract use, `_mm_extract_epi64`, the second argument selects with block of 64 bits to return, `0` for the lower 64 bits and `1` for the upper 64 bits:

use std::arch::x86_64::*;
let e0: i64 = unsafe { _mm_extract_epi64(x, 0) };
let e1: i64 = unsafe { _mm_extract_epi64(x, 1) };

As of March 2024, `u64x2` is still nightly while the code above is stable.
